### PR TITLE
Working battle royale multiplayer & ecstatic game background

### PR DESCRIPTION
**Pull request by msoechting**

### CODE REVIEW

File: src/tetris/brick/block.ts [hunk 1]
```diff
@@ -31,11 +31,9 @@ export default class Block {
 
 	//region public methods
 	public destroy(): void {
-		if (this._brick) {
-			this._brick.blocks.splice(this._brick.blocks.indexOf(this), 1);
-		}
-		this._sprite.setVisible(false);
+		//this._sprite.setVisible(false);
```
Comment (julianweise) [inline src/tetris/brick/block.ts]: Is this still necessary?

File: src/tetris/field/field.ts [hunk 2]
```diff
@@ -66,6 +82,10 @@ export default class Field {
 	//endregion
 
 	//region public methods
+	public addMoveBonusPoints(rows: number = 1) {
```
Comment (julianweise) [inline src/tetris/field/field.ts]: `void` is missing

File: src/tetris/field/field.ts [hunk 3]
```diff
@@ -126,6 +146,20 @@ export default class Field {
 			|| position.x >= this.width
 			|| position.y >= this.height);
 	}
+
+	public reset(): void {
+		for (const blockRow of this._blockRows)  {
```
Comment (julianweise) [inline src/tetris/field/field.ts]: nesting level isn't that nice.
An alternative would be to use `forEach` and/or split it to multiple functions.
Splitting it up would as a sideline increase readability,

File: src/tetris/field/field.ts [hunk 4]
```diff
@@ -167,7 +202,9 @@ export default class Field {
 		this._activeBrick = this._brickFactory.newBrick(this);
 		this._nextActiveBrickDrop = time + this._activeBrickDropInterval;
 		if (this._activeBrick.checkIfStuck()) {
+			console.log('brick stuck')
```
Comment (julianweise) [inline src/tetris/field/field.ts]: `console.log` really necessary here?

File: src/tetris/field/field.ts [hunk 5]
```diff
@@ -223,7 +260,7 @@ export default class Field {
 	}
 
 	private _increaseScore(deletedRows: number): void {
-		this._score += deletedRows * 100;
+		this._score += deletedRows * 1000;
```
Comment (julianweise) [inline src/tetris/field/field.ts]: Magic number. I would suggest to put it into a separate config or, at least, introduce a constant, defined at the top of this file.

File: src/tetris/game.ts [hunk 6]
```diff
@@ -86,7 +86,7 @@ export default class Game extends Phaser.Game {
 		this._startOfMatchSubscribers = [];
 		this._profiler = new Profiler(this);
 		this._biasEngine = new BiasEngine(this._profiler);
-		this._networkingClient = new NetworkingClient();
+		this._networkingClient = new NetworkingClient();
```
Comment (julianweise) [inline src/tetris/game.ts]: remove unnecessary whitespace

File: src/tetris/interfaces/Match.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import MatchPlayer from "tetris/interfaces/MatchPlayer";
+
+export default interface Match {
+	id: number;
+	players: MatchPlayer[];
+	startTime: string; // need to call Date.parse first, then it's a Date
```
Comment (julianweise) [inline src/tetris/interfaces/Match.ts]: Why not using `Date` in the first place?

File: src/tetris/networking/networkingClient.ts [hunk 8]
```diff
@@ -9,8 +9,8 @@ export default class NetworkingClient {
 	//endregion
 
   //region public methods
-  public emit(event: string, args) {
-    this._socket.emit(event, args);
+  public emit(event: string, args: any, acknowledgement: Function = null) {
```
Comment (julianweise) [inline src/tetris/networking/networkingClient.ts]: `any` is usually way to generic. Would suggest to make it more explicit.

Comment (julianweise) [reply src/tetris/networking/networkingClient.ts]: `void` is missing

File: src/tetris/scene/menuScene.ts [hunk 9]
```diff
@@ -40,9 +51,14 @@ export default class MenuScene extends Phaser.Scene {
 	private _optionsButton: TextButton;
 	private readonly _game: Game;
 	private _pipeline: Phaser.Renderer.WebGL.Pipelines.TextureTintPipeline;
+	private _localPlayerName: String;
 	//endregion
 
 	//region private methods
+	private async _setLocalPlayerName(): Promise<void> {
+		this._localPlayerName = (document.querySelector('#playername') as any).value;
```
Comment (julianweise) [inline src/tetris/scene/menuScene.ts]: Why not use `as HTMLElement` or `as HTMLInputElement`?

File: src/tetris/scene/menuScene.ts [hunk 10]
```diff
@@ -66,10 +82,42 @@ export default class MenuScene extends Phaser.Scene {
...
+	private _leaveMatchmaking(): void {
+		this._game.networkingClient.emit("leaveMatchmaking", {});
+	}
+
+	private _registerNetworkEvents(): void {
+		this._game.networkingClient.receive("matchmakingUpdate", this._updateMatchmakingInfo.bind(this));
+		this._game.networkingClient.receive("matchReady", this._joinMatch.bind(this));
+	}
+
+	private _updateMatchmakingInfo(matchmakingUpdate: MatchmakingInfo) {
```
Comment (julianweise) [inline src/tetris/scene/menuScene.ts]: `void` is missing

File: src/tetris/scene/playScene.ts [hunk 11]
```diff
@@ -137,53 +92,111 @@ export default class PlayScene extends Phaser.Scene {
...
+	private _startMatch(): void {
+		if (this._localPlayerField.fieldState === FieldState.Loss) {
+			return;
+		}
+		console.log('starting match');
+		this.adjustBackgroundParameters(0.25, [0.7, 0.7, 0.7]);
+		this._localPlayerField.reset();
+	}
+
+	private _updateMatch(match: Match) {
```
Comment (julianweise) [inline src/tetris/scene/playScene.ts]: `void` is missing

File: src/tetris/scene/playScene.ts [hunk 12]
```diff
@@ -137,53 +92,111 @@ export default class PlayScene extends Phaser.Scene {
...
+		console.log(match);
```
Comment (julianweise) [inline src/tetris/scene/playScene.ts]: Necessary?

File: src/tetris/scene/playScene.ts [hunk 13]
```diff
@@ -137,53 +92,111 @@ export default class PlayScene extends Phaser.Scene {
...
+		this._match = match;
+		const matchHasStarted = Date.now() > Date.parse(match.startTime);
+		if (!this._startTimerStarted && !matchHasStarted) {
+			this._startTimerStarted = true;
+			setTimeout(this._startMatch.bind(this), Date.parse(match.startTime) - Date.now(), {});
+		}
+		for (const [index, player] of match.players.entries()) {
+			if (player.socketId === this._localSocketId) {
```
Comment (julianweise) [inline src/tetris/scene/playScene.ts]: Nesting-level and readability.
I would suggest to split it up to multiple private functions.
That would increase readability as well.

File: src/tetris/scene/playScene.ts [hunk 14]
```diff
@@ -195,51 +208,52 @@ export default class PlayScene extends Phaser.Scene {
...
-		this._pipeline.setFloat1('uTime', time / 800);
+		this._rainbowPipeline.setFloat1('uTime', time / 800);
+		this._backgroundGraphicsPipeline.setFloat1('uTime', time / 1200);
 	}
 
-	private _startMatch(): void {
-		this._match = new Match();
-		this._game.handleStartOfMatch(this._match);
-	}
+	private _pushMultiplayerUpdate() {
```
Comment (julianweise) [inline src/tetris/scene/playScene.ts]: `void` is missing

File: src/tetris/ui/scoreboardWidget.ts [hunk 15]
```diff
@@ -108,58 +79,107 @@ export default class ScoreboardWidget {
 	//endregion
 
     //region private methods
-    private _clearScoreboard(): void {
-        for (let index = 0; index < this.ranks.length; index++) {
-            this.ranks[index].destroy();
-            this.names[index].destroy();
-            this.scores[index].destroy();
+    private _createTextObjects(): void {
+        for (let index = 0; index < DISPLAYED_PLAYERS; index++) {
```
Comment (julianweise) [inline src/tetris/ui/scoreboardWidget.ts]: There are several functions looping somehow over `DISPLAYED_PLAYERS`.
I would suggest to introduce a helper function that loops over it and executes custom payload passed in via arguments

File: src/tetris/field/field.ts (hunk 3 above)
Comment (friedow) [reply src/tetris/field/field.ts]: I don't think that this is necessary here, because the code reads straightforward. It just iterates over all blocks, how would you split this?

File: src/tetris/interfaces/Match.ts (hunk 7 above)
Comment (friedow) [reply src/tetris/interfaces/Match.ts]: Because this is the interface for objects that got serialized over the network via JSON where everything is a string, a number or an Object.

File: src/tetris/scene/menuScene.ts (hunk 9 above)
Comment (friedow) [reply src/tetris/scene/menuScene.ts]: Nice suggestion, will do :)

File: src/tetris/scene/playScene.ts (hunk 13 above)
Comment (friedow) [reply src/tetris/scene/playScene.ts]: I just refactored this, would you have another look, please?

File: src/tetris/ui/scoreboardWidget.ts (hunk 15 above)
Comment (friedow) [reply src/tetris/ui/scoreboardWidget.ts]: Could you give an example of how this could work?

File: src/tetris/field/field.ts [hunk 16]
```diff
@@ -36,13 +36,29 @@ export default class Field {
...
+	public set fieldStateChanged(fieldState: boolean) {
+		this._fieldStateChanged = fieldState;
+	}
+
 	public get fieldState(): FieldState {
 		return this._fieldState;
 	}
+	
+	public set fieldState(fieldState: FieldState) {
+		this._fieldState = fieldState;
```
[the remark this answers is not in the document]
Comment (friedow) [reply src/tetris/field/field.ts]: @msoechting would you have a look, please?

File: src/tetris/field/field.ts [hunk 17]
```diff
@@ -66,6 +82,10 @@ export default class Field {
 	//endregion
 
 	//region public methods
+	public addMoveBonusPoints(rows: number = 1) {
+		this._score += 3 * rows;
```
[the remark this answers is not in the document]
Comment (friedow) [reply src/tetris/field/field.ts]: I moved this into a private constant, which should be better to deal with later on.

File: src/tetris/interfaces/MatchPlayer.ts [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+
+export default interface MatchPlayer {
+	placement: number;
+	displayName: string;
+	socketId: string;
+    points: number;
+	connectionStatus: ConnectionStatus;
+	scoreboardStatus: ScoreboardStatus;
+	playStatus: PlayStatus;
+	field: Object;
```
[the remark this answers is not in the document]
Comment (friedow) [reply src/tetris/interfaces/MatchPlayer.ts]: @msoechting would you have a look, please?

File: src/tetris/scene/menuScene.ts [hunk 19]
```diff
@@ -66,10 +82,42 @@ export default class MenuScene extends Phaser.Scene {
...
+	private _joinMatchmaking(): void {
+		this._game.networkingClient.emit("joinMatchmaking", {});
+	}
+
+	private _leaveMatchmaking(): void {
+		this._game.networkingClient.emit("leaveMatchmaking", {});
+	}
+
+	private _registerNetworkEvents(): void {
+		this._game.networkingClient.receive("matchmakingUpdate", this._updateMatchmakingInfo.bind(this));
```
[the remark this answers is not in the document]
Comment (friedow) [reply src/tetris/scene/menuScene.ts]: I don't see how this should work, because this is event based not asynchronous.

File: src/tetris/scene/menuScene.ts [hunk 20]
```diff
@@ -66,10 +82,42 @@ export default class MenuScene extends Phaser.Scene {
...
+		this._game.networkingClient.receive("matchReady", this._joinMatch.bind(this));
+	}
+
+	private _updateMatchmakingInfo(matchmakingUpdate: MatchmakingInfo) {
+		// TODO: Update Matchmaking Widget
+		console.log("matchmakingUpdate: " + JSON.stringify(matchmakingUpdate));
+	}
+
+	private _joinMatch(match: Match): void {
+		this._game.networkingClient.emit("joinMatch", { matchId: match.id, displayName: this._localPlayerName }, (result: JoinResult) => {
```
[the remark this answers is not in the document]
Comment (friedow) [reply src/tetris/scene/menuScene.ts]: I'm not sure whether or not this logic should be moved to the networkingClient. What do you think @msoechting?

File: src/tetris/scene/playScene.ts [hunk 21]
```diff
@@ -13,111 +13,66 @@ import CountdownWidget from "tetris/ui/countdownWidget";
...
 
 		// Update UI widgets
 		this._scoreWidget.update(this._localPlayerField.score.toString());
-		this._countdownWidget.update(30 - (time / 1000 % 30), 30 );
 
-		if (this._localPlayerField.fieldState == FieldState.Playing && this._localPlayerField.blockStateChanged) {
-			this._localPlayerField.blockStateChanged = false;
-			this._game.networkingClient.emit("fieldUpdate", { fieldState: this._localPlayerField.serializedBlockState});
+		const matchStart = new Date(this._match.startTime).valueOf() / 1000;
+		const currentTime = new Date().valueOf() / 1000;
```
[the remark this answers is not in the document]
Comment (friedow) [reply src/tetris/scene/playScene.ts]: Nice suggestion, thanks 👍.

File: src/tetris/ui/scoreboardWidget.ts [hunk 22]
```diff
@@ -108,58 +79,107 @@ export default class ScoreboardWidget {
...
+            if (index >= players.length) {
+                break;
+            }
+            this.ranks[index].setText(players[index].placement.toString());
+            this.names[index].setText(players[index].displayName);
+            this.scores[index].setText(players[index].points.toString());
+
+            // Set text color
+            if (players[index].playStatus === PlayStatus.Eliminated) {
```
[the remark this answers is not in the document]
Comment (friedow) [reply src/tetris/ui/scoreboardWidget.ts]: I refactored this, would you have another look, please?

File: src/tetris/field/field.ts (hunk 3 above)
Comment (julianweise) [reply src/tetris/field/field.ts]: `this._blockRows.forEach(blockRow => blockRow.forEach(block => block.destroy()));`

File: src/tetris/ui/scoreboardWidget.ts (hunk 15 above)
Comment (julianweise) [reply src/tetris/ui/scoreboardWidget.ts]:
```
private _applyToAllDisplayedPlayers(payLoad: (index: number) => void) {
    for (let index = 0; index < DISPLAYED_PLAYERS; index++) {
        payload(index);
    }
}

private _someFunction() {
    this._applyToAllDisplayedPlayers(index => {
        // some stuff to do
    });
}

File: src/tetris/field/field.ts (hunk 3 above)
Comment (friedow) [reply src/tetris/field/field.ts]: That's exactly what I wanted to avoid. I think this would rather decrease the readability.

File: src/tetris/networking/networkingClient.ts [hunk 23]
```diff
@@ -9,19 +9,19 @@ export default class NetworkingClient {
...
     this._socket.on(event, callback);
   }
 
-  public connect() {
+  public connect(): void {
     this._socket.connect();
   }
 
-  public getSocketId(): string {
+  public get socketId(): string {
```
[the remark this answers is not in the document]
Comment (msoechting) [reply src/tetris/networking/networkingClient.ts]: Done

File: src/tetris/scene/playScene.ts [hunk 24]
```diff
@@ -13,111 +13,66 @@ import CountdownWidget from "tetris/ui/countdownWidget";
...
+		const preGame = currentTime < matchStart;		
+		if (!preGame && this._match.nextElimination) {
+			const nextElimination = new Date(this._match.nextElimination).valueOf() / 1000;
+			this._countdownWidget.update(Math.max(0, nextElimination - currentTime), Math.max(nextElimination - matchStart, matchStart - currentTime), preGame);
+		} else {
+			this._countdownWidget.update(Math.max(0, matchStart - currentTime), 30, preGame);
```
[the remark this answers is not in the document]
Comment (msoechting) [reply src/tetris/scene/playScene.ts]: Fixed this, the constant was only temporary

File: src/tetris/scene/playScene.ts [hunk 25]
```diff
@@ -13,111 +13,66 @@ import CountdownWidget from "tetris/ui/countdownWidget";
...
-];
-
 export default class PlayScene extends Phaser.Scene {
 
 	//region public members
 	//endregion
 
 	//region public methods
 	public preload(): void {
+		this.load.image("noise", "./assets/images/noise.png");
```
[the remark this answers is not in the document]
Comment (msoechting) [reply src/tetris/scene/playScene.ts]: Done

File: src/tetris/interfaces/MatchPlayer.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (msoechting) [reply src/tetris/interfaces/MatchPlayer.ts]: Done

File: src/tetris/field/field.ts (hunk 3 above)
Comment (julianweise) [reply src/tetris/field/field.ts]: okay
